drivers/eeprom/eeprom_fm25w256.c

```diff
@@ -0,0 +1,209 @@
+#include <drivers/eeprom.h>
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+#include <logging/log.h>
+
+#define DT_DRV_COMPAT cypress_fm25w256
+
+// 256-Kbit ferroelectric random access memory logically organized as 32K × 8 bit
+#define SIZE KB(32)
+#define WORD_SIZE 8
+
+LOG_MODULE_REGISTER(fm25w256, CONFIG_EEPROM_LOG_LEVEL);
+
+// Opcode Commands
+enum {
+	OPCODE_WREN = 0b00000110,
+	OPCODE_WRDI = 0b00000100,
+	OPCODE_RDSR = 0b00000101,
+	OPCODE_WRSR = 0b00000001,
+	OPCODE_READ = 0b00000011,
+	OPCODE_WRITE = 0b00000010,
+};
+
+struct config {
+	struct spi_dt_spec spi_spec;
+	struct gpio_dt_spec wp_spec;
+	struct gpio_dt_spec hold_spec;
+	bool read_only;
+};
+
+static int configure_pin(const struct gpio_dt_spec *spec)
+{
+	if (!spec->port) {
+		return -ENODEV;
+	}
+
+	if (!device_is_ready(spec->port)) {
+		return -EACCES;
+	}
+
+	return gpio_pin_configure_dt(spec, GPIO_OUTPUT_INACTIVE);
+}
+
+static int init(const struct device *dev)
+{
+	const struct config *config = dev->config;
+	int ret;
+
+	ret = configure_pin(&config->wp_spec);
+	switch (ret) {
+	case 0:
+		break;
+	case -ENODEV:
+		LOG_DBG("Unable to configure write protect pin: Device not ready (%d)", ret);
+		break;
+	default:
+		LOG_ERR("Unable to configure write protect pin: %d", ret);
+		return ret;
+	}
+
+	ret = configure_pin(&config->hold_spec);
+	switch (ret) {
+	case 0:
+		break;
+	case -ENODEV:
+		LOG_DBG("Unable to configure HOLD pin: Device not ready (%d)", ret);
+		break;
+	default:
+		LOG_ERR("Unable to configure HOLD pin: %d", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int write_enable(const struct device *dev)
+{
+	uint8_t opcode[] = { OPCODE_WREN };
+	const struct spi_buf buf[] = {
+		{
+			.buf = opcode,
+			.len = sizeof(opcode),
+		},
+	};
+
+	struct spi_buf_set tx = {
+		.buffers = buf,
+		.count = 1,
+	};
+
+	const struct config *config = dev->config;
+
+	return spi_write_dt(&config->spi_spec, &tx);
+}
+
+static int memory_operation(const struct device *dev, off_t offset, void *buf, size_t buf_len,
+			    bool is_write)
+{
+	int ret;
+
+	uint8_t opcode[] = {
+		is_write ? OPCODE_WRITE : OPCODE_READ,
+	};
+
+	uint8_t mem_addr[] = {
+		(offset >> 8) & 0xff,
+		offset & 0xff,
+	};
+
+	const struct spi_buf spi_buf[] = {
+		{
+			.buf = opcode,
+			.len = sizeof(opcode),
+		},
+		{
+			.buf = mem_addr,
+			.len = sizeof(mem_addr),
+		},
+		{
+			.buf = buf,
+			.len = buf_len,
+		},
+	};
+
+	struct spi_buf_set tx_buf_set = {
+		.buffers = spi_buf,
+		.count = is_write ? ARRAY_SIZE(spi_buf) : ARRAY_SIZE(spi_buf) - 1,
+	};
+
+	struct spi_buf_set rx_buf_set = {
+		.buffers = spi_buf,
+		.count = ARRAY_SIZE(spi_buf),
+	};
+
+	const struct config *config = dev->config;
+
+	if (is_write && config->read_only) {
+		LOG_ERR("Device is configured read only");
+		return -EACCES;
+	}
+
+	if (offset > SIZE) {
+		LOG_ERR("Offset out of range");
+		return -ENOMEM;
+	}
+
+	if (offset + buf_len > SIZE) {
+		LOG_ERR("Memory roll over");
+		return -EACCES;
```

Comment:
The AT24 driver returns `-EINVAL` when attempting to write past the size, we should use the same value here; an error here is result of invalid offset or buf_len, so `-EINVAL` seems to be appropriate.